src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts

```diff
@@ -407,67 +414,100 @@ export class ObjectExplorerService implements IObjectExplorerService {
+							 */
+							if (resultMap.size !== allProviders.length) {
+								const missingProviders = allProviders.filter(p => !resultMap.has(p.providerId));
+								this.logService.warn(`${session.sessionId}: Node expansion timed out for node ${node.nodePath} for providers ${missingProviders.map(p => p.providerId).join(', ')}`);
+								this._notificationService.error(nls.localize('nodeExpansionTimeout', "Node expansion timed out for node {0} for providers {1}", node.nodePath, missingProviders.map(p => p.providerId).join(', ')));
+							}
+							resolveExpansion();
+						}, expansionTimeout * 1000);
+
+						self._sessions[session.sessionId!].nodes[node.nodePath].expandEmitter.event((expandResult: NodeExpandInfoWithProviderId) => {
+							if (expandResult && expandResult.providerId) {
+								this.logService.trace(`${session.sessionId}: Received expand result for node ${node.nodePath} from provider ${expandResult.providerId}`);
+								resultMap.set(expandResult.providerId, expandResult);
+								// If we got an error result back then send error our error event
+								// We only do this for the MSSQL provider
+								if (expandResult.errorMessage) {
+									const errorType = expandResult.errorMessage.indexOf('Object Explorer task didn\'t complete') !== -1 ? 'Timeout' : 'Other';
+									// For folders send the actual name of the folder (since the nodeTypeId isn't useful in this case and the names are controlled by us)
+									const nodeType = expandResult.providerId === mssqlProviderName && node.nodeTypeId === NodeType.Folder ? node.label : node.nodeTypeId;
+									this._telemetryService.createErrorEvent(TelemetryKeys.TelemetryView.Shell, TelemetryKeys.TelemetryError.ObjectExplorerExpandError, undefined, errorType)
+										.withAdditionalProperties({
+											nodeType,
+											providerId: expandResult.providerId
+										}).send();
 								}
-							}, error => {
-								reject(error);
-							});
+
+							} else {
+								this.logService.error('OE provider returns empty result or providerId');
+							}
+
+							// When get all responses from all providers, merge results
+							if (resultMap.size === allProviders.length) {
+								resolveExpansion();
+							}
 						});
+						if (newRequest) {
+							allProviders.forEach(provider => {
+								self.callExpandOrRefreshFromProvider(provider, {
+									sessionId: session.sessionId!,
+									nodePath: node.nodePath,
+									securityToken: session.securityToken
+								}, refresh).then(isExpanding => {
+									if (!isExpanding) {
+										// The provider stated it's not going to expand the node, therefore do not need to track when merging results
+										let emptyResult: azdata.ObjectExplorerExpandInfo = {
+											errorMessage: undefined,
+											nodePath: node.nodePath,
+											nodes: [],
+											sessionId: session.sessionId
+										};
+										resultMap.set(provider.providerId, emptyResult);
```

Comment:
Should do the resultMap size check here too in case it's the last one - otherwise we'll wait the full timeout duration before checking